frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -760,6 +760,29 @@ TypeInference::assignment(const IR::Node* errorPosition, const IR::Type* destTyp
     return sourceExpression;
 }
 
+const IR::Node* TypeInference::postorder(IR::Annotation* annotation) {
+    auto checkAnnotation = [this] (const IR::Expression* e) {
+        if (!isCompileTimeConstant(e))
+            typeError("%1%: structured annotation must be compile-time constant values", e);
+        auto t = getType(e);
+        if (!t->is<IR::Type_InfInt>() &&
+            !t->is<IR::Type_String>() &&
+            !t->is<IR::Type_Boolean>())
+            typeError("%1%: illegal type for structured annotation; must be int, string or bool", e);
```

Comment:
Do we want to be more specific about the `int` type in the error message, i.e. specify infinite-precision?